File: gfapy/line/edge/link/link.py
from ..common.alignment_type import AlignmentType
from ..common.from_to import FromTo
from ..gfa1.to_gfa2 import ToGFA2 as GFA1_ToGFA2
from ..gfa1.references import References as GFA1_References
from ..gfa1.oriented_segments import OrientedSegments
from ..gfa1.alignment_type import AlignmentType as GFA1_AlignmentType
from ..gfa1.other import Other
from .canonical import Canonical
from .complement import Complement
from .equivalence import Equivalence
from .references import References as Link_References
from .to_gfa2 import ToGFA2 as Link_ToGFA2
from ..edge import Edge

class Link(Link_ToGFA2, GFA1_ToGFA2, Link_References, Equivalence, Complement, \
    Canonical, Other, GFA1_AlignmentType, OrientedSegments, GFA1_References, \
    AlignmentType, FromTo, Edge):
  """A link line (L) of a GFA1 file.

  Note:
    from_segment and to_segment are used instead of from/to
    as from is not a valid method name in Python and "to" alone
    potentially clashes with the tag namespace.
  """

  RECORD_TYPE = "L"
  POSFIELDS = ["from_segment", "from_orient", "to_segment", "to_orient",
               "overlap"]
  PREDEFINED_TAGS = ["MQ", "NM", "RC", "FC", "KC", "ID"]
  DATATYPE = {
    "from_segment" : "segment_name_gfa1",
    "from_orient" : "orientation",
    "to_segment" : "segment_name_gfa1",
    "to_orient" : "orientation",
    "overlap" : "alignment_gfa1",
    "MQ" : "i",
    "NM" : "i",
    "RC" : "i",
    "FC" : "i",
    "KC" : "i",
    "ID" : "Z",
  }
  NAME_FIELD = "ID"
  REFERENCE_FIELDS = ["from_segment", "to_segment"]
  BACKREFERENCE_RELATED_FIELDS = ["to_orient", "from_orient", "overlap"]
  DEPENDENT_LINES = ["paths"]

Link._apply_definitions()

File: lib/rgfa/line/edge/link.rb
# A link connects two segments, or a segment to itself.
# @tested_in api_positionals, api_references_edges_gfa1
class RGFA::Line::Edge::Link < RGFA::Line::Edge

  RECORD_TYPE = :L
  POSFIELDS = [:from, :from_orient, :to, :to_orient, :overlap]
  PREDEFINED_TAGS = [:MQ, :NM, :RC, :FC, :KC]
  FIELD_ALIAS = {}
  DATATYPE = {
     :from => :segment_name_gfa1,
     :from_orient => :orientation,
     :to => :segment_name_gfa1,
     :to_orient => :orientation,
     :overlap => :alignment_gfa1,
     :MQ => :i,
     :NM => :i,
     :RC => :i,
     :FC => :i,
     :KC => :i,
  }
  NAME_FIELD = nil
  STORAGE_KEY = nil
  REFERENCE_FIELDS = [:from, :to]
  BACKREFERENCE_RELATED_FIELDS = [:to_orient, :from_orient, :overlap]
  DEPENDENT_LINES = [:paths]
  OTHER_REFERENCES = []

  apply_definitions

end

require_relative "common/alignment_type"
require_relative "common/from_to"
require_relative "gfa1/to_gfa2"
require_relative "gfa1/references"
require_relative "gfa1/oriented_segments"
require_relative "gfa1/alignment_type"
require_relative "gfa1/other"
require_relative "link/canonical"
require_relative "link/complement"
require_relative "link/equivalence"
require_relative "link/references"
require_relative "link/to_gfa2"

class RGFA::Line::Edge::Link
  include RGFA::Line::Edge::Common::FromTo
  include RGFA::Line::Edge::Common::AlignmentType
  include RGFA::Line::Edge::GFA1::ToGFA2
  include RGFA::Line::Edge::GFA1::References
  include RGFA::Line::Edge::GFA1::OrientedSegments
  include RGFA::Line::Edge::GFA1::AlignmentType
  include RGFA::Line::Edge::GFA1::Other
  include RGFA::Line::Edge::Link::Canonical
  include RGFA::Line::Edge::Link::Complement
  include RGFA::Line::Edge::Link::Equivalence
  include RGFA::Line::Edge::Link::References
  include RGFA::Line::Edge::Link::ToGFA2
end

